src/network/networkd-dhcp4.c

```diff
@@ -138,10 +138,13 @@ static int link_set_dhcp_routes(Link *link) {
         }
 
         n = sd_dhcp_lease_get_routes(link->dhcp_lease, &static_routes);
-        if (n == -ENODATA)
-                return 0;
         if (n < 0)
-                return log_link_warning_errno(link, n, "DHCP error: could not get routes: %m");
+                log_link_warning_errno(link, n, "DHCP error: could not get routes: %m");
```

Comment:
It's not really an error for DHCP not to send a route. It could be just interface is unroutable to the outside....

Reply:
This pr is about dhcp not sending any routes. So what I am missing here. It's waring message.

Reply:
If sd_dhcp_lease_get_routes returns -ENODATA, this will still print an error. You have to exclude this case like above:
`if (n < 0 && n != -ENODATA)`

Reply:
I think we should log the warning message .

Reply:
This comment is about this change and the one in line 90.

If I got the intend of this PR right this is about DHCP responses that do not carry a `ROUTER` attribute aka a default gateway.

IMHO logging this situation is definitely required. I am just unsure if `warning` is the right level here. `warning` causes me to think that there might be something wrong within `networkd` but it was able to continue vs this is a (valid) edge case that networkd did handle, but you might be wondering why there is no default route.

Wouldn't `log_link_info_errno` be better suited in that case? In any case the logged message could probably made more descriptive.

For the first message in Line 90 logging beased on one of the two return codes (-EINVAL, -ENODATA) could probably be something like
```C
switch (n) {
 case -ENODATA: log_info(…, "DHCP {nfo: No routes received from DHCP server: %m"); break ;
 case -EINVAL:
 default:
   log_{warrning,error}(…, "DHCP {warning|error}: could not get routes: %m"); break ;
}
```

I have had many frustrating hours with software that just logged errors like `Subsystem: error: ${errno}` where I had to start reading the source code to actually get why or what is going wrong. Getting this right in networkd would be awesome :-)

Reply:
I get your point . We should be more more informative about what actually going wrong . will update and please feel free to comment on these issues which help us debugging more user friendly. Thank you.

Reply:
Yeah, if the function returns `-ENODATA`, we should show messages at most in info level, I think.
What do you think about applying the logic similar to one for `sd_dhcp_lease_get_router()`?

Reply:
No this not a normal case we don't know the routes. That means it is not routble. Info is just a information we can ignore. I don't think we can ignore this.

Reply:
Hmm, I may misunderstand the code or what the routes here are. But on my laptop, I got
```
wlp2s0: DHCP error: could not get routes: No data available
```
however I can connect internet. I thought that even if we do not get routes from DHCP, getting default gateway is sufficient to connect outside. Is it correct, or not?

Reply:
Yes that is correct. But we can't really keep the log in info level.

Reply:
Perhaps "DHCP server did not provide any additional routes" would be a nicer message...

Reply:
@ssahani Then, I still do not understand why you stick to warn here. I think we should warn if DHCP server do not provide either default gateway or additional routes. What do you think about the following?
```
if (n == -ENODATA)
    log_link_info(link, "DHCP: No routes received from DHCP server: %m");
else if (n < 0)
    log_link_warning_errno(link, n, "DHCP error: could not get routes: %m");
if (r < 0 && n < 0)
    log_link_warning(link, "DHCP error: No routes received from DHCP server: %m");
```

Reply:
Agree this is more informative.